### ISSUE TITLE
stack broken on kretprobe

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#101]
May be broken on uretprobe as well. For example:

```
# ./src/bpftrace -e 'kprobe:__alloc_fd { @entry[stack] = count(); } kretprobe:__alloc_fd { @return[stack] = count(); }'
Attaching 2 probes...
^C

@entry[
__alloc_fd+1
__do_pipe_flags+87
do_pipe2+45
__x64_sys_pipe+16
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 1
@entry[
__alloc_fd+1
do_sys_open+273
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 1
@entry[
__alloc_fd+1
__do_pipe_flags+73
do_pipe2+45
__x64_sys_pipe+16
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 1
@entry[
__alloc_fd+1
do_sys_open+273
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 1
@entry[
__alloc_fd+1
do_sys_open+273
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 4
@entry[
__alloc_fd+1
do_sys_open+273
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 9
@entry[
__alloc_fd+1
do_sys_open+273
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 13
@entry[
__alloc_fd+1
__sys_socket+114
__x64_sys_socket+22
do_syscall_64+85
entry_SYSCALL_64_after_hwframe+68
]: 14


@return[
kretprobe_trampoline+0
]: 1
@return[
kretprobe_trampoline+0
]: 2
@return[
kretprobe_trampoline+0
]: 4
@return[
kretprobe_trampoline+0
]: 6
@return[
kretprobe_trampoline+0
]: 9
@return[
kretprobe_trampoline+0
]: 27
```

The stack shouldn't change between the kprobe (entry) and kretprobe (return). But on the return we just get this kretprobe_trampoline.

**Comment by cyphar** [comment]
I've also seen this if you manually instrument `kprobe_events` -- `$stack` on a `kretprobe` results in a garbage stack (with only the entry `[unknown/kretprobe'd]` -- which is `kretprobe_trampoline`). Is there any particular place that you would recommend looking at if I wanted to help fix this issue (I've needed the ability to get a stack trace from a `kretprobe` in the past so this is quite useful to me).

**Comment by cyphar** [comment]
I looked into this a bit more. It's actually quite conceptually easy to do with the current `kretprobe` design -- all you have to do is save the stack trace inside `pre_handler_kretprobe` (`kretprobes` are actually `kprobes` that set up a `kretprobe` once they get triggered -- which means that we can just get a stack trace during the code which sets up the `kretprobe` and then we can return that if a user asks us for a stack trace during the `kretprobe`). So getting a stack trace on function entry is very doable.

The problem is how we get that information to our BPF tracer. Getting it to `ftrace` and all the other kernel tracers (which all appear to be implementing somewhat similar concepts but differently -- presumably because they were all developed separately or because they have strong performance reasons for doing so) appears to be possible -- but getting it to the BPF program is not as easy.

Effectively we need `bpf_get_stackid` (I believe) to sample from a different stack than the current one -- which means that we have to get a reference to `kretprobe_instance`. We get this for free in `kretprobe_perf_func`. But the BPF program doesn't have access to that -- `trace_call_bpf` doesn't take nor give that information. We'd have to modify what we pass in `ctx` to allow that -- but given that it's all done with `void *` I'm quite scared of breaking lots of things by touching it.

Also, `bpf_get_stackid` uses the `perf_events` stack trace code which appears to be somewhat different from the `/proc/self/stack` stack trace code -- they both use the same `orc` unwinder on `x86` but they have different interfaces for some reason. It's not clear to me which we should use for `kretprobe_instance` given that `ftrace` also has its own stack-getting code...

**Comment by brendangregg** [comment]
Thanks for digging in @cyphar . Sounds like a kernel change, and you know more about fixing this right now than anyone, so if you can I'd try it out! I'm guessing the patch would be submitted via BPF https://github.com/torvalds/linux/blob/master/MAINTAINERS#L2810 , depending on what changes.

**Comment by cyphar** [comment]
Okay, I have a working patch. You can see it in [my tree](https://github.com/cyphar/linux/tree/kretprobe-stack). The basic idea is that we stash away the stack trace on `kprobe` entry and then use it within the tracing code if the tracing code is executing from within a `kretprobe`.

**Comment by cyphar** [comment]
Saving the stack trace was NACKed by upstream. However, I've figured out how to make the unwinders produce reasonable stack traces... but unfortunately this doesn't produce identical traces to syscall entry (my original idea was flawed but with the new code could be useful).

So yeah, this requires a bit more work.

**Comment by wenjianhn** [comment]
I think the kernel issue has been fixed by https://lore.kernel.org/all/163163049242.489837.11970969750993364293.stgit@devnote2/T/#u

Linux v5.16 contains the fix.

For older kernels like v5.4, `CONFIG_UNWINDER_FRAME_POINTER=y` is likely to be a workaround.

**Comment by jordalgo** [comment]
Closing this as it was fixed in 5.16 and we have a workaround for earlier kernels (mentioned above).
